File: medium/155_min_stack/min_stack.py
class MinStack:

    def __init__(self):
        self.values = []
        self.min_with = []
        self.top_index = -1

    def push(self, val: int) -> None:
        self.values.append(val)
        if len(self.min_with) == 0:
            self.min_with.append(val)
        else:
            self.min_with.append(
                min(val, self.min_with[max(self.top_index, 0)]))
        self.top_index += 1

    def pop(self) -> None:
        self.values.pop()
        self.min_with.pop()
        self.top_index -= 1

    def top(self) -> int:
        return self.values[self.top_index]

    def getMin(self) -> int:
        return self.min_with[self.top_index]

File: leetcode/medium/155_min_stack/min_stack.cpp
#include <vector>

class MinStack {
public:
  MinStack() {}

  void push(int val) {
    stack.push_back(val);
    if (min_stack.size() == 0 || val <= min_stack.back()) {
      min_stack.push_back(val);
    }
  }

  void pop() {
    if (stack.back() == min_stack.back()) {
      min_stack.pop_back();
    }
    stack.pop_back();
  }

  int top() { return stack.back(); }

  int getMin() { return min_stack.back(); }

private:
  std::vector<int> stack;
  std::vector<int> min_stack;
};
